.github/workflows/test.yml

```diff
@@ -7,127 +7,163 @@ on:
-          # Run the tests with gotestsum
-          env CGO_ENABLED=${{ matrix.cgo_enabled }} ./gotestsum -- -v -count=10 -shuffle=on ./...
-
-  linux-other:
-    runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        arch: ["arm64"]
-        cgo_enabled: [ "0", "1" ] # test it compiles with and without the cgo
-      fail-fast: false
-    steps:
-      - uses: actions/checkout@v3
-      - name: Go modules cache
-        uses: actions/cache@v3
-        with:
-          path: ~/go/pkg/mod
-          key: go-pkg-mod-${{ matrix.arch }}-${{ hashFiles('**/go.sum') }}
-          restore-keys: go-pkg-mod-${{ matrix.arch }} go-pkg-mod-
       - name: Set up QEMU
         uses: docker/setup-qemu-action@v2
         with:
           platforms: ${{ matrix.arch }}
-      - run: docker run --platform=linux/${{ matrix.arch }} -v${HOME}/go/pkg/mod:/root/go/pkg/mod -v $PWD:$PWD -w $PWD -eCGO_ENABLED=${{ matrix.cgo_enabled }} -eDD_APPSEC_WAF_TIMEOUT=$DD_APPSEC_WAF_TIMEOUT golang go test -v -count=10 -shuffle=on ./...
+      - name: Create container
+        id: container
+        run: |-
+          docker run --name gha-${{ github.run_id }} --rm -di                   \
+            --platform="linux/${{ matrix.arch }}"                               \
+            -v "${HOME}/go/pkg/mod:/go/pkg/mod"                                 \
+            -v "$PWD:$PWD"                                                      \
+            -w "$PWD"                                                           \
+            -eCGO_ENABLED="${{ matrix.cgo-enabled }}"                           \
+            -eDD_APPSEC_WAF_TIMEOUT="${DD_APPSEC_WAF_TIMEOUT}"                  \
+            -eGODEBUG="${{ matrix.go-debug }}"                                  \
+            -eGOEXPERIMENT="${{ matrix.go-experiment }}"                        \
+            -eGOMODCACHE="/go/pkg/mod"                                          \
+            "${{ format(matrix.image, matrix.go-version) }}"
+      - name: Install alpine requirements
+        if: endsWith(matrix.image, '-alpine') && matrix.cgo-enabled == '1'
+        run: |-
+          docker exec -i gha-${{ github.run_id }}                               \
+            apk add gcc musl-dev libc6-compat
+      - name: Install AmazonLinux 2 requirements
+        if: matrix.image == 'amazonlinux:2'
+        run: |-
+          docker exec -i gha-${{ github.run_id }}                               \
+            yum install -y golang
+      - name: Install gotestsum
+        run: |-
+          docker exec -i gha-${{ github.run_id }}                               \
+            go install gotest.tools/gotestsum@latest
+      - name: go test
+        run: |-
+          docker exec -i gha-${{ github.run_id }}                               \
+            go run gotest.tools/gotestsum@latest --                             \
+              -v -count=10 -shuffle=on -tags='${{ matrix.go-tags }}'             \
+              ./...
```

Comment:
I usually prefer a multi-stage Dockerfile where the final container image only contains the produced binary and its dependencies, without all of the dev tools. This allows us to find setup limitations/issues.

OTOH, it would be valid to say this test's role is just to execute the tests on the given containers, regardless of the setup/onboarding use-cases.

Reply:
But what I also think is that setup cases should be tested in the appsec test app https://github.com/DataDog/appsec-go-test-app/blob/main/Dockerfile at the very least

Reply:
I'd need to figure out how to pre-build test binaries and correctly execute them then... but I'm not against the idea. We'd then also want bare-metal tests to have separated build & run jobs (where run only has the test binary available).